scripts/util/charmhub.py

```diff
@@ -0,0 +1,198 @@
+        archs = item.get_archs()
+
+        for revision_matrix in self.data.values():
+            if revision_matrix.get_bases() != bases or revision_matrix.get_archs() != archs:
+                return False
+            
+            for base in bases:
+                for arch in archs:
+                    if item.get(arch, base) and not revision_matrix.get(arch, base):
+                            return False
+                    
+        return True
+
+    def get_bases(self):
+        try:
+            item: RevisionMatrix = random.choice(list(self.data.values())) # nosec
+            return item.get_bases()
+        except StopIteration:
+            return set()
+        
+    def get_archs(self):
+        try:
+            item: RevisionMatrix = random.choice(list(self.data.values())) # nosec
+            return item.get_archs()
+        except StopIteration:
+            return set()
+    
+    def get_revisions(self, arch, base):
+        revisions = {}
+
+        for charm in self.data.keys():
+            revisions[f"{charm.replace("-", "_")}_revision"] = self.data[charm].get(arch, base)
+
+        return revisions
+
+    def get_version(self, arch, base):
+        
+        charms = sorted(self.data.keys())
+        if not charms:
+            return None
+
+        version = "k8s-operator"
+        for charm in charms:
+            revision_matrix = self.data[charm]
+            if not revision_matrix:
+                return None
+                
+            revision = revision_matrix.get(arch, base)
+            if not revision:
+                return None
+
+            version += f"-{charm}-{revision}"
+
+        return version
+
+class RevisionMatrix:
+    """
+    For each tuple of (name, channel, arch, base) there is a unique charm artifact
+    in Charmhub. RevisionMatrix is a matrix of (arch, base) revisions, if any, for
+    a specific (name, channel) tuple.
+    Rows of the matrix correspond to different architectures.
+    Columns of the matrix correspond to different bases.
+    """
+
+    def __init__(self):
+        self.data: defaultdict[str, dict[str, str]] = defaultdict(dict)
+
+    def set(self, arch, base, revision):
+        self.data[arch][base] = revision
+
+    def get_archs(self):
+        return set(self.data.keys())
+
+    def get_bases(self):
+        bases: set[str] = set()
+        for base in self.data.values():
+            bases.update(base.keys())
+        return bases
+
+    def get(self, arch, base):
+        return self.data.get(arch, {}).get(base, None)
+
+    def remove_arch(self, arch):
+        if arch in self.data:
+            del self.data[arch]
+
+    def remove_base(self, base):
+        for arch in list(self.data):
+            if base in self.data[arch]:
+                del self.data[arch][base]
+
+    def __eq__(self, other):
+        if not isinstance(other, RevisionMatrix):
+            return NotImplemented
+        return dict(self.data) == dict(other.data)
```

Comment:
unnecessary type checking.  `NotImplemented` is kind of fruitless


```suggestion
        return self.data == other.data
```

Reply:
done, thank you